### PR TITLE
feat (react-components): Make a general multipurpose tree view [BND3D-4691]

### PR DESCRIPTION
**Pull request by nilscognite**
#### Type of change

Feat  

#### Jira ticket :blue_book:

https://cognitedata.atlassian.net/browse/BND3D-4691

## Description :pencil:

This is the first tree view PR.

I think I have made it flexible for all of you. Basically you will only write the function for loading the nodes for lazy loading, and what happen when you check and select the nodes, and of coarse build up the initial tree.
It is super simple to use, since I have mocked up a few different trees with different strategies for checkboxes, selection and so on (see movie). You can also optionally use icons with colors, I have very good experience with this in user interfaces for 3D.

There is a README.txt in src\components\Architecture\TreeView\README.txt for how to use the tree.

**TODO:** It will need some styling for the colors, and I don't have control of the size. Need also to add in arrow buttons events if wanted.

## How has this been tested? :mag:

yarn storybook
open TreeView folder
Click on Main

## Test instructions :information_source:

You will find  3 different trees, with different config. Play with them

1. Dependent on checkboxes & Single select & Icons
2. Independent checkboxes  & Single Select & Icons
3. No checkboxes & No icons & Multi select

[screen-recorder-wed-oct-23-2024-11-13-59.webm](https://github.com/user-attachments/assets/d9d77a7f-5fda-4459-bba1-f607cc14eda6)

## Checklist :ballot_box_with_check:

- [x] I am proud of this feature.
- [x] I have performed a self-review of my own code.
- [x] I have added PR type (Feat, Bug, Chore, Test, Docs, Style, Refactor) to the PR title.
- [x] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [x] I have commented my code in hard-to-understand areas.
- [x] I have made corresponding changes to the public documentation.
- [ ] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [ ] I have refactored the code for readability to the best of my ability.
- [ ] I have checked that my changes do not introduce regressions in the public documentation.
- [ ] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [ ] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [ ] I have added TSDoc to any public facing changes.
- [ ] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [ ] I have noted down and am currently tracking any technical debt introduced in this PR.
- [x] I have thoroughly thought about the architecture of this implementation.
- [x] I have asked peers to test this feature.

### CODE REVIEW

**Review by danpriori — CHANGES_REQUESTED** [review]
It looks good to me overall. Nice work!

I really liked the structure of the classes and how it organized the logic to load the nodes, children and siblings. It is relatively easy to understand and I believe it is easy to scale.

I have some comments that I find interesting to address, such as the callbacks that the application can use to receive the list of selected TreeNodes, and also the change of some files and functions to facilitate future changes and improvements in the organization and structure of some components and other comments for reflection and questions that I have about it.

We can discuss some points for changes in the future, because I don't think that these things that I raised are blockers now. Perhaps the crucial point at the moment is my questions about the callbacks that the application would use to receive the Array of TreeNodes.

I tested a little and the performance seems to be good. I forced the tree generator to create more sub levels and apparently the performance remains reasonable.

I'm gonna test a little more trying to do some stress test as well. But looks like stable so far.

File: react-components/src/components/Architecture/TreeView/TreeView.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/*!
+ * Copyright 2023 Cognite AS
+ */
+
+/* eslint-disable react/prop-types */
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/components/Architecture/TreeView/TreeView.tsx:5]: Fixed

File: react-components/src/components/Architecture/TreeView/TreeView.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+export const TreeView = (props: TreeViewProps): ReactElement => {
+  const id = 'treeView';
+  const nodes = getChildrenAsArray(props.root, props.loadNodes, false);
+  if (nodes === undefined) {
+    return <></>;
+  }
+  return (
+    <div
+      // ref={htmlRef}
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/components/Architecture/TreeView/TreeView.tsx:21]: Fixed

File: react-components/src/components/Architecture/TreeView/TreeViewNode.tsx [hunk 3]
```diff
@@ -0,0 +1,376 @@
+/*!
+ * Copyright 2023 Cognite AS
+ */
+
+/* eslint-disable react/prop-types */
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/components/Architecture/TreeView/TreeViewNode.tsx:5]: We need this for some reason

File: react-components/src/components/Architecture/TreeView/TreeViewNode.tsx [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+
+  return (
+    <div>
+      <div
+        style={{
+          flexDirection: 'row',
+          display: 'flex',
+          gap: gapBetweenItems,
+          marginTop: gapBetweenItems,
+          marginLeft: level * gapToChildren + 'px'
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/components/Architecture/TreeView/TreeViewNode.tsx:84]: Ok

File: react-components/stories/TreeView.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        if (i % 7 === 0) child.iconColor = 'red';
+        array.push(child);
+      }
+      resolve(array);
+    }, 2000)
+  );
+  return await promise;
+}
+
+function createTreeMock(lazyLoading: boolean): ITreeNode {
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/stories/TreeView.stories.tsx:130]: Done

File: react-components/src/components/Architecture/TreeView/TreeViewNode.tsx [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+}: {
+  node: ITreeNode;
+  level: number;
+  props: TreeViewProps;
+}): ReactElement => {
+  const [hoverOverTextOrIcon, setHoverOverTextOrIcon] = useState(false);
+
+  // This force to update the component when the node changes
+  // See https://coreui.io/blog/how-to-force-a-react-component-to-re-render/
+  const [, forceUpdate] = useReducer((x) => x + 1, 0);
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/components/Architecture/TreeView/TreeViewNode.tsx:53]: It work pretty well and saved me from making tons of code lines.

File: react-components/src/components/Architecture/TreeView/TreeViewNode.tsx [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+}): ReactElement => {
+  const label = props.loadingLabel ?? LOADING_LABEL;
+  const gapBetweenItems = (props.gapBetweenItems ?? GAP_BETWEEN_ITEMS) + 'px';
+  const gapToChildren = props.gapToChildren ?? GAP_TO_CHILDREN;
+  return (
+    <div
+      style={{
+        gap: gapBetweenItems,
+        marginTop: gapBetweenItems,
+        marginLeft: (level + 1) * gapToChildren + 'px'
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/components/Architecture/TreeView/TreeViewNode.tsx:324]: Fixed

**Review by nilscognite — COMMENTED** [review]
I have fixed some of the easy things. The rest I have to discuss with you

File: react-components/src/components/Architecture/TreeView/TreeView.tsx [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      id={id}
+      style={{
+        backgroundColor: props.backgroundColor ?? BACKGROUND_COLOR
+      }}>
+      {nodes.map((node, index) => (
+        <TreeViewNode node={node} key={index} level={0} props={props} />
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/components/Architecture/TreeView/TreeView.tsx:24]: I added a getSelectedNodes and getCheckedNodes method on TreeNode.

File: react-components/src/components/Architecture/TreeView/TreeViewNode.tsx [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+  isHoverOver: boolean
+): string | undefined {
+  if (!node.isParent) {
+    return 'transparent';
+  }
+  if (isHoverOver) {
+    return props.hoverCaretColor ?? HOVER_CARET_COLOR;
+  }
+  return props.caretColor ?? CARET_COLOR;
+}
```
[the remark this answers is not in the document]
Comment (nilscognite) [reply react-components/src/components/Architecture/TreeView/TreeViewNode.tsx]: done

**Review by danpriori — APPROVED** [review]
Based on our discussion and with the review of the new changes, LGTM. Dont forget to remove the unnecessary commits from the another PR.

Nice work!

**Pull request merged**